src/h5web/dataset-visualizer/vis/MatrixVis.tsx

```diff
@@ -13,34 +14,162 @@ interface Props {
   data: HDF5Value;
 }
 
+type CellProps = Omit<GridChildComponentProps, 'data'>;
+
+interface MatrixProps {
+  children: CellRendererFunc;
+  width: number;
+  height: number;
+}
+
+interface IndexCellProps {
+  index: number;
+  style: React.CSSProperties;
+}
+
+type CellRendererFunc = (props: GridChildComponentProps) => JSX.Element;
+
 function MatrixVis(props: Props): JSX.Element {
   const { dims, data } = props;
+  const gridDimensions = {
+    rowHeight: 32,
+    rowCount: dims[0] + 1,
+    columnWidth: 116,
+    columnCount: (dims.length === 2 ? dims[1] : 1) + 1,
+  };
+  const GridSettings = createContext(gridDimensions);
+
+  const CellRenderer: CellRendererFunc = (
+    cellRendererProps: GridChildComponentProps
+  ) => {
+    const { data: itemData, rowIndex, columnIndex, style } = cellRendererProps;
+    const { NormalCellFunc } = itemData;
+    // Disable rendering of index columns (done by the innerElementType)
+    if (rowIndex * columnIndex === 0) {
+      return <></>;
+    }
+    return (
+      <NormalCellFunc
+        rowIndex={rowIndex - 1}
+        columnIndex={columnIndex - 1}
+        style={style}
+      />
+    );
+  };
 
   const accessor: AccessorFunc =
     dims.length === 1 ? row => data[row] : (row, col) => data[row][col];
 
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  function ValueCell(cellProps: CellProps): JSX.Element {
+    const { rowIndex: row, columnIndex: col, style } = cellProps;
+    return (
+      <div
+        className={styles.cell}
+        style={style}
+        data-bg={(col + row) % 2 === 1 ? '' : undefined}
+      >
+        {numeral(accessor(row, col)).format('0.000e+0')}
+      </div>
+    );
+  }
+
+  function IndexCell(indexCellProps: IndexCellProps): JSX.Element {
+    const { index, style } = indexCellProps;
+    return (
+      <div className={styles.indexCell} style={style}>
+        {index >= 0 && index}
+      </div>
+    );
+  }
+
+  function IndexColumn(): JSX.Element {
+    const { columnWidth, rowHeight, rowCount } = useContext(GridSettings);
+    return (
+      <div
+        className={`${styles.sticky} ${styles.indexColumn}`}
```

Comment:
I tried to separate the `index` CSS from the `sticky` CSS. Perhaps can we implement a toggle sticky ?

Reply:
The sticky state is dynamic now, so there are a couple of ways to implement this a bit better than by conditionally concatenating classes:

- using a lib like `classnames`;
- adding a ternary and two classes: `indexColumn` and `stickyIndexColumn`, and composing `sticky` from the latter in CSS;
- moving the sticky state to a `data` attribute: `data-sticky={sticky || undefined}` and using an attribute selector: `.indexColumn[data-attribute]`.

The last solution is the cleanest in my opinion. The only downside it that it increases the specificity of the selector, which could be bad in some cases.